pkg/pgmodel/querier/query_builder.go

```diff
@@ -427,6 +427,30 @@ func calledByTimestamp(path []parser.Node) bool {
 }
 
 var vectorSelectorExtensionRange = semver.MustParseRange(">= 0.1.3-beta")
+var rateIncreaseExtensionRange = semver.MustParseRange(">= 0.1.3-beta")
+
+func callAggregator(hints *storage.SelectHints, funcName string) *aggregators {
+	queryStart := hints.Start + hints.Range
+	queryEnd := hints.End
+	stepDuration := time.Second
+	rangeDuration := time.Duration(hints.Range) * time.Millisecond
+
+	if hints.Step > 0 {
+		stepDuration = time.Duration(hints.Step) * time.Millisecond
+	} else {
+		if queryStart != queryEnd {
+			panic("query start should equal query end")
+		}
+	}
+	qf := aggregators{
+		timeClause:  "ARRAY(SELECT generate_series($%d::timestamptz, $%d::timestamptz, $%d))",
+		timeParams:  []interface{}{model.Time(queryStart).Time(), model.Time(queryEnd).Time(), stepDuration},
+		valueClause: "prom_" + funcName + "($%d, $%d,$%d, $%d, time, value)",
+		valueParams: []interface{}{model.Time(hints.Start).Time(), model.Time(queryEnd).Time(), int64(stepDuration.Milliseconds()), int64(rangeDuration.Milliseconds())},
+		unOrdered:   false,
+	}
+	return &qf
+}
 
 /* The path is the list of ancestors (direct parent last) returned node is the most-ancestral node processed by the pushdown */
 func getAggregators(hints *storage.SelectHints, qh *QueryHints, path []parser.Node) (*aggregators, parser.Node, error) {
```

Comment:
This function needs a bit of refactoring. 

I suggest handling the various checks that overlap for all pushdowns first and returning the no optimization path as soon as possible based on them. Things like is the extension installed or if there is a subquery should be checked as soon as possible.

Should make the rest of the code much more easier to read and maintain.